From e5b0a249245c6067d6c1ae66c9589f97773cc0f3 Mon Sep 17 00:00:00 2001
From: Yiwei Zhang <zzyiwei@chromium.org>
Date: Thu, 19 Mar 2026 17:09:57 -0700
Subject: [PATCH] server: allow cleanly spawn subprocess

This helps skip:
- render_worker_jail_create
- render_server_run

Part-of: <https://gitlab.freedesktop.org/virgl/virglrenderer/-/merge_requests/1607>
---
 server/render_server.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/server/render_server.c b/server/render_server.c
index b9589bfb..efa73133 100644
--- a/server/render_server.c
+++ b/server/render_server.c
@@ -62,6 +62,9 @@ render_server_poll(UNUSED struct render_server *srv,
 static bool
 render_server_run(struct render_server *srv)
 {
+   if (srv->state == RENDER_SERVER_STATE_SUBPROCESS)
+      return true;
+
    struct render_client *client = srv->client;
 
    struct pollfd poll_fds[RENDER_SERVER_POLL_COUNT];
@@ -219,6 +222,13 @@ render_server_init(struct render_server *srv,
    if (!render_server_parse_options(srv, argc, argv))
       return false;
 
+   if (srv->context_args->ctx_fd >= 0) {
+      assert(srv->client_fd < 0);
+      srv->context_args->valid = true;
+      srv->state = RENDER_SERVER_STATE_SUBPROCESS;
+      return true;
+   }
+
    enum render_worker_jail_seccomp_filter seccomp_filter =
       RENDER_WORKER_JAIL_SECCOMP_NONE;
    const char *seccomp_path = NULL;
-- 
2.55.0

